samcli/lib/sync/sync_flow.py

```diff
@@ -93,7 +103,15 @@ def set_up(self) -> None:
         self._session = Session(profile_name=self._deploy_context.profile, region_name=self._deploy_context.region)
 
     def _boto_client(self, client_name: str):
-        return get_boto_client_provider_from_session_with_config(cast(Session, self._session))(client_name)
+        default_retry_config = get_default_retry_config()
+        if not default_retry_config:
+            LOG.debug("Creating boto client (%s) with user's retry config", client_name)
+            return get_boto_client_provider_from_session_with_config(cast(Session, self._session))(client_name)
```

Comment:
This was due to `self._session` object is defined as `Optional[Session]` since it is instantiated later via another method in the class. I've created `_get_session()` method which will lazy instantiate the object and will also remove the `Optional` type in its return.